wasm-demo/src/lib.rs

```diff
@@ -249,4 +259,90 @@ impl WorldState {
+        let line_index = self.analysis.file_line_index(self.file_id).unwrap();
+
+        let struct_nodes = match self.analysis.file_structure(self.file_id) {
+            Ok(struct_nodes) => struct_nodes,
+            _ => return JsValue::NULL,
+        };
+        let mut parents: Vec<(DocumentSymbol, Option<usize>)> = Vec::new();
+
+        for symbol in struct_nodes {
+            let doc_symbol = DocumentSymbol {
+                name: symbol.label.clone(),
+                detail: symbol.detail.unwrap_or(symbol.label),
+                kind: symbol.kind.conv(),
+                range: symbol.node_range.conv_with(&line_index),
+                children: None,
+                tags: [if symbol.deprecated { SymbolTag::Deprecated } else { SymbolTag::None }],
+                containerName: None,
+                selectionRange: symbol.navigation_range.conv_with(&line_index),
+            };
+            parents.push((doc_symbol, symbol.parent));
+        }
+        let mut res = Vec::new();
+        while let Some((node, parent)) = parents.pop() {
+            match parent {
+                None => res.push(node),
+                Some(i) => {
+                    let children = &mut parents[i].0.children;
+                    if children.is_none() {
+                        *children = Some(Vec::new());
+                    }
+                    children.as_mut().unwrap().push(node);
+                }
+            }
+        }
+
+        serde_wasm_bindgen::to_value(&res).unwrap()
+    }
+
+    pub fn type_formatting(&self, line_number: u32, column: u32, ch: char) -> JsValue {
+        log::warn!("type_formatting");
+        let line_index = self.analysis.file_line_index(self.file_id).unwrap();
+
+        let mut pos = Position { line_number, column }.conv_with((&line_index, self.file_id));
+        pos.offset = pos.offset - TextUnit::of_char('.');
+
+        let edit = match ch {
+            '=' => self.analysis.on_eq_typed(pos),
+            '.' => self.analysis.on_dot_typed(pos),
+            _ => return JsValue::NULL,
+        };
+
+        let edit = match edit {
+            Ok(Some(mut it)) => it.source_file_edits.pop().unwrap(),
+            _ => return JsValue::NULL,
+        };
+
+        let change: Vec<TextEdit> = edit.edit.conv_with(&line_index);
+        serde_wasm_bindgen::to_value(&change).unwrap()
+    }
+
+    pub fn folding_ranges(&self) -> JsValue {
+        log::warn!("folding_ranges");
+        let line_index = self.analysis.file_line_index(self.file_id).unwrap();
+        return if let Ok(folds) = self.analysis.folding_ranges(self.file_id) {
+            let res: Vec<_> = folds.into_iter().map(|fold| fold.conv_with(&line_index)).collect();
+            serde_wasm_bindgen::to_value(&res).unwrap()
+        } else {
+            JsValue::NULL
+        };
+    }
+
+    pub fn goto_implementation(&self, line_number: u32, column: u32) -> JsValue {
+        log::warn!("goto_implementation");
```

Comment:
I would expect to see info! here, but apparently that’s too slow, right? How is logging controlled on wasm? not via an env var, I assume?

Reply:
Right, that's because the logging doesn't happen through `env_logger` which doesn't support wasm. The `console_log` logger used here is a bit more bare-bones, and only supports one top-level `LogLevel` filter in the init function.
I put the filter on Warn because logging to the browser's console through wasm is insanely slow. Any lower level would start pushing a lot of Info messages and delay all actions. This way none of the messages from within rust-analyzer make it through though...
Definitely room for improvement on the logging story here 😄

Reply:
That is, we can't filter by *crate*, right?